File: manual/data/cpp/tests/test_missingno_bar.cpp
#include <gtest/gtest.h>
#include "../src/function_missingno_bar.cpp"

class TestMissingnoBar : public ::testing::Test {
protected:
    MissingnoBar missingno_bar;
};

TEST_F(TestMissingnoBar, TestCompleteData) {
    std::string csv_data = "A,B,C\n1,2,3\n4,5,6\n7,8,9";
    EXPECT_EQ(missingno_bar.visualize_missing_data(csv_data), "Missing data visualization generated.");
}

TEST_F(TestMissingnoBar, TestMissingValues) {
    std::string csv_data = "A,B,C\n1,2,\n,5,6\n7,,9";
    EXPECT_EQ(missingno_bar.visualize_missing_data(csv_data), "Missing data visualization generated.");
}

TEST_F(TestMissingnoBar, TestAllMissing) {
    std::string csv_data = "A,B,C\n,,\n,,\n,,";
    EXPECT_EQ(missingno_bar.visualize_missing_data(csv_data), "Missing data visualization generated.");
}

TEST_F(TestMissingnoBar, TestEmptyCSV) {
    std::string csv_data = "";
    EXPECT_THROW(missingno_bar.visualize_missing_data(csv_data), std::invalid_argument);
}

TEST_F(TestMissingnoBar, TestIncorrectColumnCount) {
    std::string csv_data = "A,B,C\n1,2\n4,5,6,7\n8,9";
    EXPECT_THROW(missingno_bar.visualize_missing_data(csv_data), std::invalid_argument);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: manual/data/java/tests/test_missingno_bar.java
package com.example;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;
import static org.junit.jupiter.api.Assertions.*;

public class test_missingno_bar {

    private function_missingno_bar missingnoBar;

    @BeforeEach
    public void setUp() {
        missingnoBar = new function_missingno_bar();
    }

    @Test
    public void test_complete_data() {
        String csv_data = "A,B,C\n1,2,3\n4,5,6\n7,8,9";
        String result = missingnoBar.visualize_missing_data(csv_data);
        assertEquals("Missing data visualization generated.", result);
    }

    @Test
    public void test_missing_values() {
        String csv_data = "A,B,C\n1,2,\n,5,6\n7,,9";
        String result = missingnoBar.visualize_missing_data(csv_data);
        assertEquals("Missing data visualization generated.", result);
    }

    @Test
    public void test_all_missing() {
        String csv_data = "A,B,C\n,,\n,,\n,,";
        String result = missingnoBar.visualize_missing_data(csv_data);
        assertEquals("Missing data visualization generated.", result);
    }

    @Test
    public void test_empty_csv() {
        String csv_data = "";
        assertThrows(IllegalArgumentException.class, () -> {
            missingnoBar.visualize_missing_data(csv_data);
        });
    }

    @Test
    public void test_incorrect_column_count() {
        String csv_data = "A,B,C\n1,2\n4,5,6,7\n8,9";
        assertThrows(IllegalArgumentException.class, () -> {
            missingnoBar.visualize_missing_data(csv_data);
        });
    }
}